doozerlib/dblib.py

```diff
@@ -18,47 +70,330 @@ def __init__(self, runtime, environment, dry_run=False):
+            payload_columns.add(column)
+
+        if not self._table_column_cache[table_name]:
+
+            # if the table's column cache is not updates, load the cache and return
+            # the table columns
+
+            self.runtime.logger.info("Updating table [{}] column cache.".format(table_name))
+
+            cursor = self.connection.cursor()
+            cursor.execute(f"show columns from {table_name}")
+
+            for data in cursor:
+                existing_columns.add(data[0])
+                self._table_column_cache[table_name][data[0]] = True
+        else:
+
+            # otherwise if table column cache found, return cache value
+
+            self.runtime.logger.info("Found table [{}] column cache.".format(table_name))
+            existing_columns = set(self._table_column_cache[table_name])
+
+        # get the difference column, non-existing table columns which are present in payload
+        missing_columns = list(payload_columns.difference(existing_columns))
+
+        for index, column in enumerate(missing_columns):
+            self.runtime.logger.info("Missing column {}. {}".format(index + 1, column))
+            return_columns.append((column, self.identify_column_type(attr_payload[column])))
+        return return_columns
+
+    def handle_missing_table(self, table_name):
+
+        """
+        :param table_name
+        This method creates table, if it doesn't exist.
+        """
+
+        table_exist_status = self.check_table_exist(table_name)
+
+        if not table_exist_status:
+            self.runtime.logger.info("Table [{}] not found in database [{}]."
+                                     .format(table_name, self.db))
+            cursor = self.connection.cursor()
+            cursor.execute(f'create table {table_name}(log_{table_name}_id bigint auto_increment primary key)')
+            cursor.close()
+            self._table_column_cache[table_name] = dict()
+            self._table_column_cache[table_name]["log_" + table_name + "_id"] = True
+            self.runtime.logger.info("Successfully create table [{}] in database [{}]."
+                                     .format(table_name, self.db))
+
+    def handle_missing_columns(self, payload, table_name):
+
+        """
+        :param payload
+        :param table_name
+
+        This method creates missing columns, which are present in payload but not in table.
+        """
+
+        missing_columns = self.get_missing_columns(payload, table_name)
+
+        if missing_columns:
+            cursor = self.connection.cursor()
+
+            for column in missing_columns:
+                missing_column = column[0]
+                column_type = column[1][0]
+                cursor.execute(f"alter table {table_name} add column `{missing_column}` {column_type}")
+                self._table_column_cache[table_name][missing_column] = True
+
+                self.runtime.logger.info("Added new column [{}] of identified type [{}] to table [{}] "
```

Comment:
It's interesting how we use a mix of f-strings and format() for formatting here, any particular reason? should we stick to one of them?

Reply:
We wrote a lot of code in python 2 when f-strings weren't available. Existing code and copy-pasta would likely get that.

Personally I greatly prefer f-strings.